### PR TITLE
TIP-1353: Validate localizable/scopable values

### PR DESCRIPTION
**Pull request by mmetayer**
**Description (for Contributor and Core Developer)**

The goal of this PR is to add validation for scopable/localizable value, meaning:
- at the value creation level, we now only validate the structure (for instance if the attribute is localizable, the value should have a locale code, if the attribute is a multi select, the data should be an array of strings, etc...), but not the data itself
- the validation of the data is now performed at validation stage, before persisting the data. In this PR we added 2 validators which check the consistency of the provided channel and locale codes

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | OK
| Added integration tests           | -
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                            '%channel_code%' => $localizableValue->getScopeCode(),
+                            '%invalid_locale%' => $localizableValue->getLocaleCode(),
+                        ]
+                    )->atPath(sprintf('[%s]', $key))->addViolation();
+
+                    continue;
+                }
+            }
+
+            $attribute = $this->attributeRepository->findOneByIdentifier($localizableValue->getAttributeCode());
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php]: Lru cache query

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ScopableValuesValidator extends ConstraintValidator
+{
+    /** @var ChannelRepositoryInterface */
+    private $channelRepository;
+
+    /** @var null|array */
+    private $indexedChannelCodes = null;
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php]: avoid stateful service, it's very hard to test and it should not be done at this layer (it's infra). We should use it at the infrastructure level. It does not matter to do 3 query to get the channels. I would use the channel cache repository.

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    '%channel%' => $channelCode,
+                ])->addViolation();
+            }
+        }
+    }
+
+    private function getIndexedChannelCodes(): array
+    {
+        if ($this->indexedChannelCodes === null) {
+            $channels = $this->channelRepository->findAll();
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php]: This optimization is usless as the number of channels is very low.

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php (hunk 2 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValuesValidator.php]: yeah, that's what I was thinking, and channels are most probably already cached at this point

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php (hunk 1 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php]: we don't have access to locale specific / available locale codes in `Akeneo\Pim\Structure\Component\Query\PublicApi\AttributeType\Attribute`. 
I'll do it in another PR or this one will get hard to read

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                        '%invalid_locale%' => $localizableValue->getLocaleCode(),
+                    ]
+                )->atPath(sprintf('[%s]', $key))->addViolation();
+
+                continue;
+            }
+
+            if ($localizableValue->isScopable()) {
+                $channel = $this->channelRepository->findOneByIdentifier($localizableValue->getScopeCode());
+                if (null !== $channel && !in_array($localizableValue->getLocaleCode(), $channel->getLocaleCodes())) {
```
[the remark this answers is not in the document]
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/LocalizableValuesValidator.php]: You should put this comment directly in the code, it's not obvious.

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValues.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ScopableValues extends Constraint
+{
+    public $unknownScopeMessage = 'Attribute "%attribute_code%" expects an existing scope, "%channel%" given.';
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValues.php]: expect an existing channel, no ?

File: tests/features/pim/enrichment/product/validation/localizable_values.feature [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  @acceptance-back
+  Scenario: Providing a locale not bound to the channel should raise an error
+    When a product is created with values:
+      | attribute   | data       | scope     | locale  |
+      | description | My product | ecommerce | fr_FR   |
+    Then the error 'Attribute "description" expects a valid locale, "fr_FR" is not bound to channel "ecommerce".' is raised
+
+  @acceptance-back
+  Scenario: Providing a locale part of a locale specififc attribute's available locales should not rais anb error
```
Comment (pierallard) [inline tests/features/pim/enrichment/product/validation/localizable_values.feature]: Are you drunk ?

File: tests/features/pim/enrichment/product/validation/localizable_values.feature [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    When a product is created with values:
+      | attribute | data | locale |
+      | gdpr      | test | fr_FR  |
+    Then no error is raised
+
+  @acceptance-back
+  Scenario: Providing a locale which is not part of a locale specififc attribute's available locales should not rais anb error
```
Comment (pierallard) [inline tests/features/pim/enrichment/product/validation/localizable_values.feature]: Same here

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValues.php (hunk 5 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ScopableValues.php]: unfortunately this error message is used a lot, and is not only thrown by the validator, so a lot of tests use it. It was easier to keep the original error message than to update all tests/services/... which use this message

File: tests/features/pim/enrichment/product/validation/localizable_values.feature (hunk 6 above)
Comment (mmetayer) [reply tests/features/pim/enrichment/product/validation/localizable_values.feature]: oops

**Pull request merged**